api/envoy/api/v2/core/address.proto

```diff
@@ -28,14 +28,17 @@ message SocketAddress {
   }
   Protocol protocol = 1 [(validate.rules).enum.defined_only = true];
   // The address for this socket. :ref:`Listeners <config_listeners>` will bind
-  // to the address or outbound connections will be made. An empty address is
-  // not allowed, specify ``0.0.0.0`` or ``::`` to bind any. It's still possible to
-  // distinguish on an address via the prefix/suffix matching in
-  // FilterChainMatch after connection. For :ref:`clusters
-  // <config_cluster_manager_cluster>`, an address may be either an IP or
-  // hostname to be resolved via DNS. If it is a hostname, :ref:`resolver_name
-  // <envoy_api_field_core.SocketAddress.resolver_name>` should be set unless default
-  // (i.e. DNS) resolution is expected.
+  // to the address. An empty address is not allowed. Specify ``0.0.0.0`` or ``::``
+  // to bind to any address. [#comment:TODO(zuercher) reinstate when implemented:
+  // It is possible to distinguish a Listener address via the prefix/suffix matching
+  // in :ref:`FilterChainMatch <envoy_api_msg_listener.FilterChainMatch>`.] When used
+  // within :ref:`BindConfig <envoy_api_msg_core.BindConfig>`, the address controls
+  // the source address of outbound connections. For :ref:`clusters
+  // <config_cluster_manager_cluster>`, the cluster type determines whether the
+  // address must be an IP (*STATIC* or *EDS* clusters) or a hostname resolved by DNS
+  // (*STRICT_DNS* or *LOGICAL_DNS* clusters). IP address resolution can be
+  // customized via :ref:`resolver_name
```

Comment:
Resolvers actually take things other than IP addresses. The address field is effectively an opaque string when used in conjunction with a resolver, e.g. it could contains `/some/path/in/a/made/up/name/space` which gets resolves down to IP.